addons/account/models/sequence_mixin.py

```diff
@@ -72,7 +72,14 @@ def _get_starting_sequence(self):
         return "00000000"
 
     def _get_highest_query(self):
-        return "SELECT {field} FROM {table} {where_string} ORDER BY {field} DESC LIMIT 1 FOR UPDATE"
+        regex = self._sequence_fixed_regex
+        regex_seq = re.sub(r"\?P<\w+>", "?:", regex.replace(r"?P<seq>", ""))  # make the seq the only matching group
+        regex_prefix = re.sub(r"\?P<\w+>", "?:", regex.replace(r"?P<prefix1>", ""))  # make prefix1 the only matching group
+        return self.env.cr.mogrify("""
+            SELECT {field} FROM {table} {where_string}
+            ORDER BY (regexp_match({field}, %s))[1] DESC, ('0' || (regexp_match({field}, %s))[1])::integer DESC
+            LIMIT 1 FOR UPDATE
+        """, (regex_prefix, regex_seq)).decode()
```

Comment:
if not regexp: don't do all this stuff??

Reply:
What about performance? If neither the `where` clause nor the `order by` are indexed, this will keep getting slower and slower. Some databases generate hundreds of thousands of invoices per month, sometimes in a single journal, and often in a series of batches at the end of the week or month.

Perhaps this could be solved by keeping the sequence number in a separate field, so indexing would be trivial. Oh wait. :angry:

Reply:
Without any filters (on the journal and the date) on the prod, it took 0.3s to execute the query, but it is **only** 3 times slower than without the regex.

Reply:
> it took 0.3s to execute the query, but it is _only_ 3 times slower than without the regex.

I don't know how you got those numbers?

Just ran a quick test and it's 200,000 times slower, due to the Seq Scan: 5s+ instead of 0.02.
And it's O(n) so it will get worse for the larger databases I was talking about. They can generate this kind of volume every 2 months instead of every 10 years.


```sql
d=# SELECT count(*) FROM account_move;
 count  
--------
 789342
(1 row)

-- WITH INDEX

d=# explain analyze SELECT id FROM account_move ORDER BY name DESC LIMIT 1;
                                                                           QUERY PLAN                                                                           
----------------------------------------------------------------------------------------------------------------------------------------------------------------
 Limit  (cost=0.42..0.65 rows=1 width=17) (actual time=0.013..0.013 rows=1 loops=1)
   ->  Index Scan Backward using account_move_name_index on account_move  (cost=0.42..176192.25 rows=775655 width=17) (actual time=0.012..0.012 rows=1 loops=1)
 Planning time: 0.086 ms
 Execution time: 0.025 ms
(4 rows)


-- WITHOUT INDEX

d=# EXPLAIN (ANALYZE, VERBOSE)
  SELECT id
    FROM account_move
ORDER BY (regexp_match(name, '^(.*?)(?:\d*)(?:\D*?)$'))[1] DESC,
         ('0' || (regexp_match(name, '^(?:.*?)(\d*)(?:\D*?)$'))[1])::integer DESC
   LIMIT 1;
                                                                                                      QUERY PLAN                                                                                                      
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Limit  (cost=182816.51..182816.51 rows=1 width=40) (actual time=5327.952..5327.953 rows=1 loops=1)
   Output: id, ((regexp_match((name)::text, '^(.*?)(?:\d*)(?:\D*?)$'::text))[1]), ((('0'::text || (regexp_match((name)::text, '^(?:.*?)(\d*)(?:\D*?)$'::text))[1]))::integer)
   ->  Sort  (cost=182816.51..184755.65 rows=775655 width=40) (actual time=5327.952..5327.952 rows=1 loops=1)
         Output: id, ((regexp_match((name)::text, '^(.*?)(?:\d*)(?:\D*?)$'::text))[1]), ((('0'::text || (regexp_match((name)::text, '^(?:.*?)(\d*)(?:\D*?)$'::text))[1]))::integer)
         Sort Key: ((regexp_match((account_move.name)::text, '^(.*?)(?:\d*)(?:\D*?)$'::text))[1]) DESC, ((('0'::text || (regexp_match((account_move.name)::text, '^(?:.*?)(\d*)(?:\D*?)$'::text))[1]))::integer) DESC
         Sort Method: top-N heapsort  Memory: 25kB
         ->  Seq Scan on public.account_move  (cost=0.00..178938.24 rows=775655 width=40) (actual time=0.045..5215.814 rows=789342 loops=1)
               Output: id, (regexp_match((name)::text, '^(.*?)(?:\d*)(?:\D*?)$'::text))[1], (('0'::text || (regexp_match((name)::text, '^(?:.*?)(\d*)(?:\D*?)$'::text))[1]))::integer
 Planning time: 0.071 ms
 Execution time: 5327.968 ms
(10 rows)
```

And the cost of the regular expressions _is_ significant because it is incurred for each record scanned. For example using only the prefix regexp cuts the time to 3.5s. Great, it's only 130,000 times slower now.

```sql
d=# EXPLAIN (ANALYZE, VERBOSE)
  SELECT id
    FROM account_move
ORDER BY (regexp_match(name, '^(.*?)(?:\d*)(?:\D*?)$'))[1] DESC 
   LIMIT 1;
                                                                 QUERY PLAN                                                                 
--------------------------------------------------------------------------------------------------------------------------------------------
 Limit  (cost=175059.96..175059.96 rows=1 width=36) (actual time=3552.507..3552.508 rows=1 loops=1)
   Output: id, ((regexp_match((name)::text, '^(.*?)(?:\d*)(?:\D*?)$'::text))[1])
   ->  Sort  (cost=175059.96..176999.10 rows=775655 width=36) (actual time=3552.506..3552.506 rows=1 loops=1)
         Output: id, ((regexp_match((name)::text, '^(.*?)(?:\d*)(?:\D*?)$'::text))[1])
         Sort Key: ((regexp_match((account_move.name)::text, '^(.*?)(?:\d*)(?:\D*?)$'::text))[1]) DESC
         Sort Method: top-N heapsort  Memory: 25kB
         ->  Seq Scan on public.account_move  (cost=0.00..171181.69 rows=775655 width=36) (actual time=0.124..3471.832 rows=789342 loops=1)
               Output: id, (regexp_match((name)::text, '^(.*?)(?:\d*)(?:\D*?)$'::text))[1]
 Planning time: 0.227 ms
 Execution time: 3552.545 ms
(10 rows)
```


This is clearly not an option unless we generate appropriate indexes automatically.

Reply:
My bad, the "3 times slower" was compared to using `ORDER BY length({field}) DESC, {field} DESC` instead of the regex, like proposed by @KangOl initially.

Reply:
> This is clearly not an option unless we generate appropriate indexes automatically.

It will be difficult as the regexp can be different per journal.

Reply:
Making a new module adding 2 columns: one with the format (like `PREFIX/{seq}/SUFFIX` and one with the number in `int` format.

The split can be done using the regex. Then we can do an `ORDER BY name_format DESC, name_seq DESC`